share/eds/file.go

```diff
@@ -0,0 +1,171 @@
+package eds
+
+import (
+	"io"
+	"os"
+
+	"golang.org/x/exp/mmap"
+
+	"github.com/celestiaorg/celestia-app/pkg/wrapper"
+	"github.com/celestiaorg/nmt"
+	"github.com/celestiaorg/rsmt2d"
+
+	"github.com/celestiaorg/celestia-node/share"
+)
+
+// File
+// * immutable
+// * versionable
+// TODO:
+//   - Cache Rows and Cols
+//   - Avoid storing constant shares, like padding
+type File struct {
+	path string
+	hdr  Header
+	fl   fileBackend
+}
+
+type fileBackend interface {
+	io.ReaderAt
+	io.Closer
+}
+
+func OpenFile(path string) (*File, error) {
+	f, err := mmap.Open(path)
+	if err != nil {
+		return nil, err
+	}
+
+	h, err := ReadHeaderAt(f, 0)
+	if err != nil {
+		return nil, err
+	}
+
+	return &File{
+		path: path,
+		hdr:  h,
+		fl:   f,
+	}, nil
+}
+
+// TODO: Allow setting features
+func CreateFile(path string, eds *rsmt2d.ExtendedDataSquare) (*File, error) {
+	f, err := os.Create(path)
+	if err != nil {
+		return nil, err
+	}
+
+	h := Header{
+		ShareSize:  uint16(len(eds.GetCell(0, 0))), // TODO: rsmt2d should expose this field
+		SquareSize: uint32(eds.Width()),
+	}
+
+	if _, err = h.WriteTo(f); err != nil {
+		return nil, err
+	}
+
+	for _, shr := range eds.Flattened() {
+		// TODO: Buffer and write as single?
+		if _, err := f.Write(shr); err != nil {
+			return nil, err
+		}
+	}
+
+	return &File{
+		path: path,
+		fl:   f,
+		hdr:  h,
+	}, err
+}
+
+func (f *File) Close() error {
+	return f.fl.Close()
+}
+
+func (f *File) Header() Header {
+	return f.hdr
+}
+
+func (f *File) Axis(idx int, _ rsmt2d.Axis) ([]share.Share, error) {
+	// TODO: Add Col support
+	shrLn := int64(f.hdr.ShareSize)
+	sqrLn := int64(f.hdr.SquareSize)
+	rwwLn := shrLn * sqrLn
+
+	offset := int64(idx)*rwwLn + HeaderSize
+	rowdata := make([]byte, rwwLn)
+	if _, err := f.fl.ReadAt(rowdata, offset); err != nil {
+		return nil, err
+	}
+
+	row := make([]share.Share, sqrLn)
+	for i := range row {
+		row[i] = rowdata[int64(i)*shrLn : (int64(i)+1)*shrLn]
+	}
+	return row, nil
+}
+
+func (f *File) Share(idx int) (share.Share, error) {
+	// TODO: Check the cache first
+	shrLn := int64(f.hdr.ShareSize)
+
+	offset := int64(idx)*shrLn + HeaderSize
+	shr := make(share.Share, shrLn)
+	if _, err := f.fl.ReadAt(shr, offset); err != nil {
+		return nil, err
+	}
+	return shr, nil
+}
+
+func (f *File) ShareWithProof(idx int, axis rsmt2d.Axis) (share.Share, nmt.Proof, error) {
```

Comment:
[feature request]: Having `AxisWithProof` will be super useful for search operations to retrieve all rows with their proofs and roots to be able to filter out for specific shares and their proofs.

Reply:
Yep, wanted to make it

Reply:
When you get the whole axis, there is no proof.